open_spiel/games/gin_rummy.cc

```diff
@@ -57,8 +57,194 @@ std::shared_ptr<const Game> Factory(const GameParameters& params) {
 
 REGISTER_SPIEL_GAME(kGameType, Factory);
 
+bool ObserverHasString(IIGObservationType iig_obs_type) {
+  return !iig_obs_type.perfect_recall;
+}
+
+bool ObserverHasTensor(IIGObservationType iig_obs_type) {
+  return !iig_obs_type.perfect_recall;
+}
+
 }  // namespace
 
+
+class GinRummyObserver : public Observer {
+ public:
+  GinRummyObserver(IIGObservationType iig_obs_type)
+      : Observer(/*has_string=*/ObserverHasString(iig_obs_type),
+                 /*has_tensor=*/ObserverHasTensor(iig_obs_type)),
+        iig_obs_type_(iig_obs_type) {}
+
+
+  void WriteTensor(const State& observed_state, int player,
+                   Allocator* allocator) const override {
+    auto& state = open_spiel::down_cast<const GinRummyState&>(observed_state);
+    SPIEL_CHECK_GE(player, 0);
+    SPIEL_CHECK_LT(player, state.num_players_);
+
+    // TODO(Michal) Should we implement perfect recall?
+    if (iig_obs_type_.perfect_recall) {
+      SpielFatalError(
+          "GinRummyObserver: tensor with perfect recall not implemented.");
+    }
+
+    // Observing player.
+    WriteObservingPlayer(state, player, allocator);
+
+    // Private hand(s).
+    if (iig_obs_type_.private_info == PrivateInfoType::kSinglePlayer) {
+      WriteSinglePlayerHand(state, player, allocator);
+    } else if (iig_obs_type_.private_info == PrivateInfoType::kAllPlayers) {
+      WriteAllPlayerHands(state, allocator);
+    }
+
+    // Public information.
+    if (iig_obs_type_.public_info) {
+      WriteKnockCard(state, allocator);
+      WriteUpcard(state, allocator);
+      WriteDiscardPile(state, allocator);
+      WriteStockSize(state, allocator);
+      WriteLayedMelds(state, allocator);
+    }
+  }
+
+  std::string StringFrom(const State& observed_state,
+                         int player) const override {
+    auto& state = open_spiel::down_cast<const GinRummyState&>(observed_state);
+    SPIEL_CHECK_GE(player, 0);
+    SPIEL_CHECK_LT(player, state.num_players_);
+
+    // TODO(Michal) Should we implement perfect recall?
```

Comment:
I think we should do it for strings, as that is the canonical way of identifiying infostates (for tabular algs etc.)

Reply:
Done, see above.